pkg/edvprovider/couchdbedvprovider/couchdbedvprovider.go

```diff
@@ -0,0 +1,176 @@
+// Since CouchDB won't let us index a field within an array (which is where encrypted indices are defined),
+// as a workaround, we create another document with a mapping of the encrypted index to the document that has it.
+func (c CouchDBEDVProvider) CreateIndexOnSingleField(store storage.Store, indexedAttributeName,
+	encryptedDocID string) error {
+	mapDocument := couchDBIndexMappingDocument{
+		IndexName:              indexedAttributeName,
+		MatchingEncryptedDocID: encryptedDocID,
+	}
+
+	documentBytes, err := json.Marshal(mapDocument)
+	if err != nil {
+		return err
+	}
+
+	err = store.Put(encryptedDocID+"_mapping_"+uuid.New().String(), documentBytes)
+	if err != nil {
+		return err
+	}
+
+	createIndexRequest := storage.CreateIndexRequest{
+		IndexStorageLocation: "EDV_EncryptedIndexesDesignDoc",
+		IndexName:            "EDV_IndexName",
+		WhatToIndex:          `{"fields": ["` + indexedField + `"]}`,
+	}
+
+	return store.CreateIndex(createIndexRequest)
+}
+
+// EncryptedIndexQuery does an EDV encrypted index query.
+// Due to the document mapping workaround described in CreateIndexOnSingleField, we first need to get
+// the "mapping document" and then use the ID we get from that to lookup the associated encrypted document.
+// Then we check that encrypted document to see if the value matches what was specified.
+func (c CouchDBEDVProvider) EncryptedIndexQuery(store storage.Store, query *models.Query) ([]string, error) {
+	idsOfDocsWithMatchingQueryIndexName, err := findDocsMatchingQueryIndexName(store, query.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	return findDocsMatchingQueryValue(idsOfDocsWithMatchingQueryIndexName, store, query)
+}
+
+func findDocsMatchingQueryIndexName(store storage.Store, queryIndexName string) (map[string]struct{}, error) {
+	itr, err := store.Query(`{
+		   "selector": {
+		       "` + indexedField + `": "` + queryIndexName + `"
+		   },
+			"use_index": ["EDV_EncryptedIndexesDesignDoc", "EDV_IndexName"]
+		}`)
+	if err != nil {
+		return nil, err
+	}
+
+	ok, err := itr.Next()
+	if err != nil {
+		return nil, err
+	}
+
+	// I really just want an array that doesn't allow for duplicates, but in golang a map is the closest I can get.
+	// The "struct{}" value is not used and should not occupy any additional space.
+	idsOfDocsWithAMatchingIndex := make(map[string]struct{})
+
+	for ok {
+		value, err := itr.Value()
+		if err != nil {
+			return nil, err
+		}
+
+		receivedCouchDBIndexMappingDocument := couchDBIndexMappingDocument{}
+
+		err = json.Unmarshal(value, &receivedCouchDBIndexMappingDocument)
+		if err != nil {
+			return nil, err
+		}
+
+		idsOfDocsWithAMatchingIndex[receivedCouchDBIndexMappingDocument.MatchingEncryptedDocID] = struct{}{}
+
+		ok, err = itr.Next()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return idsOfDocsWithAMatchingIndex, nil
+}
+
+func findDocsMatchingQueryValue(idsOfDocsWithAMatchingIndex map[string]struct{}, store storage.Store,
```

Comment:
The name, idsOfDocsWithAMatchingIndex, is a left-over from the calling function. We just want a simple variable name here, since this function doesn't care how the arg was created.

Reply:
Good catch! Done.